arcs/cetera.py

```diff
@@ -0,0 +1,120 @@
+import logging
+import requests
+from frozendict import frozendict
+
+from collect_domain_query_data import lang_filter
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Query(object):
+    @property
+    def text(self):
+        return self._text
+
+    @classmethod
+    def from_query_str(cls, q):
+        parts = q.split('=')
+        return FacetSearch(parts[0], parts[1]) if len(parts) == 2 else KeywordSearch(q)
+
+    def to_query_params(self):
+        pass
+
+
+class KeywordSearch(Query):
+    def __init__(self, text):
+        self._text = text
+
+    def to_query_params(self):
+        return {"q": self._text}
+
+    def __str__(self):
+        return self.text
+
+    __repr__ = __str__
+
+
+def prepare_facet(facet):
+    if facet == "category":
+        prepared = "categories"
+    else:
+        prepared = facet
+
+    return prepared
+
+
+class FacetSearch(Query):
+    def __init__(self, facet, text):
+        super(Query, self)
+        self._facet = facet
+        self._text = text
+
+    @property
+    def facet(self):
+        return self._facet
+
+    def to_query_params(self):
+        return {prepare_facet(self.facet): self.text}
+
+    def __str__(self):
+        return "{}={}".format(self.facet, self.text)
+
+    __repr__ = __str__
+
+
+def get_cetera_results(domain_query_pairs, cetera_host, cetera_port,
+                       num_results=10, cetera_params=None,
+                       filter_too_few_results=False):
+    """
+    Get the top n=num_results catalog search results from Cetera for each
+    (domain, query) pair in domain_query_pairs.
+
+    Args:
+        domain_query_pairs (Iterable[(str, Query)]): An iterable of domain-query pairs
+        cetera_host (str): The Cetera hostname
+        cetera_port (int): The port on which to make requests to Cetera
+        num_results (int): The number of results to fetch for each query
+        cetera_params (Dict[Any, Any]): Optional, additional query parameters for Cetera
```

Comment:
It appears that way, but did you check out Python 3 type annotations? As I mentioned previously, there is no comprehensive guide on the right way to do this, so I'm synthesizing the Google Python styleguide and the docs on the Python 3 type annotations (https://docs.python.org/3/library/typing.html). I find it awesomely readable.